src/freenet/support/MultiValueTable.java

```diff
@@ -1,230 +1,260 @@
-        synchronized (table) {
-            Vector<V> v = table.get(key);
-            return (v == null ?
-                    null :
-                    v.firstElement());
+        return getFirst(key);
+    }
+
+    /**
+     * Returns the first element for this key.
+     */
+    public V getFirst(K key) {
+        List<V> list = this.table.get(key);
+        if (list == null || list.isEmpty()) {
+            return null;
         }
+        return list.get(0);
     }
 
     public boolean containsKey(K key) {
-		synchronized (table) {
-			return table.containsKey(key);
-		}
+        return this.table.containsKey(key);
     }
 
     public boolean containsElement(K key, V value) {
-        synchronized (table) {
-            Vector<V> v = table.get(key);
-            return (v != null) && v.contains(value);
+        List<V> list = this.table.get(key);
+        if (list == null || list.isEmpty()) {
+            return false;
         }
+        return list.contains(value);
     }
 
     /**
-     * Users will have to handle synchronizing.
+     * Returns mapped value collection as {@link Enumeration}
+     *
+     * <b>Note:</b> this method is deprecated
+     * please use other {@link #getAllAsList(Object)} method variant, which provides a {@link List}
+     *
+     * @param key key mapping
      */
+    @Deprecated
     public Enumeration<V> getAll(K key) {
-    	Vector<V> v;
-		synchronized (table) {
-			v = table.get(key);
-		}
-        return (v == null ?
-                new EmptyEnumeration<V>() :
-                v.elements());
+        List<V> elements = getAllAsList(key);
+        if (elements.isEmpty()) {
+            return Collections.emptyEnumeration();
+        } else {
+            return Collections.enumeration(elements);
+        }
+    }
+
+    public List<V> getAllAsList(K key) {
+        List<V> list = this.table.get(key);
+        if (list == null || list.isEmpty()) {
+            return Collections.emptyList();
+        }
+        return Collections.unmodifiableList(list);
     }
 
     /**
      * To be used in for(x : y).
      */
     public Iterable<V> iterateAll(K key) {
-		synchronized (table) {
-			return(table.get(key));
-		}
+        return getAllAsList(key);
     }
 
     public int countAll(K key) {
-    	Vector<V> v;
-		synchronized (table) {
-			v = table.get(key);
-		}
-    	if(v != null)
-        	return v.size();
-        else
-        	return 0;
+        return getAllAsList(key).size();
     }
 
+    /**
+     * Returns mapped value collection as raw {@link Object}
+     *
+     * <b>Note:</b> this method is deprecated, this was used in previous {@code synchronized} implementation variant,
+     * please use other {@link #getAllAsList(Object)})} method variant, which provides a typed {@link List} collection.
+     *
+     * @param key key mapping
+     */
+    @Deprecated
     public Object getSync(K key) {
-		synchronized (table) {
-			return table.get(key);
-		}
+        return getAllAsList(key);
```

Comment:
This method used to return a `Vector` (or `null` if there was no value), now it always returns a `List`. There are many methods that are implemented by Vector but not by List, so this is a breaking change (even though this behaviour was undocumented, as usual).

Additionally, this now returns an unmodifiable list where it would previously allow modifications. We might as well remove this method right away - we cannot in any reasonable way simulate previous behaviour using this new implementation, and no sane plugin should use it anyways (given its entirely undocumented nature).

I checked the following plugins to be sure (listing by Plugin-Main-Class):
```
de.saces.fnplugins.Shoeshop.ShoeshopPlugin
de.saces.fnplugins.SiteToolPlugin.SiteToolPlugin
freenetsearch.Plugin
keepalive.Plugin
net.pterodactylus.sone.main.SonePlugin
org.freenetproject.freemail.FreemailPlugin
plugins.CENO.Client.CENOClient
plugins.floghelper.FlogHelper
plugins.flophelper.FlogHelper
plugins.FMS.FMSPlugin
plugins.Freereader.Freereader
plugins.JSTUN.JSTUN
plugins.KeyUtils.KeyUtilsPlugin
plugins.Library.Main
plugins.MDNSDiscovery.MDNSDiscovery
plugins.N2NChat.core.N2NChatPlugin
plugins.ShareLink.Plugin
plugins.Sharesite.Plugin
plugins.ShareWiki.Plugin
plugins.ThawIndexBrowser.ThawIndexBrowser
plugins.UPnP2.UPnP2
plugins.UPnP.UPnP
plugins.WebOfTrust.WebOfTrust
```

None of them use anything from MultiValueTable except:
```
freenet.support.MultiValueTable#<init>() // the no-args constructor
freenet.support.MultiValueTable#put(java.lang.Object,java.lang.Object)
```

@ArneBab @Bombe thoughts on removing this method?

Reply:
I changed this back  to return `Vector`.
It may be removed at any time later.